confdgnmi/src/confd_gnmi_adapter.py

```diff
@@ -58,126 +58,238 @@ def set(self, prefix, path, val):
+                    response = self.sample(
+                        start_monitoring=self.is_monitor_changes() and first_sample)
                     yield response
-                    if not streaming:
+                    first_sample = False
+                    if self.is_once():
                         break
-                elif self.subscription_stream_event_type == self.SubscriptionStreamEventType.FINISH:
+                elif event == self.SubscriptionEvent.FINISH:
                     log.debug("finishing subscription read")
                     break
-                elif self.subscription_stream_event_type == None:
-                    log.warning(
-                        "**** self.subscription_stream_event_type is None ! ****")
+                elif event == self.SubscriptionEvent.SEND_CHANGES:
+                    response = self.changes()
+                    log.debug("Sending changes")
+                    yield response
+                elif event is None:
+                    log.warning("**** event is None ! ****")
                     # TODO error
                     break
                 else:
-                    log.warning(
-                        "**** self.subscription_stream_event_type not processed ! ****")
+                    log.warning("**** event=%s not processed ! ****", event)
                     # TODO error
                     break
-                log.debug("Waiting self.subscription_stream_event_type=%s ...",
-                         self.subscription_stream_event_type)
-                self.thread_event.wait()
-                #TODO can there be a race condition? E.g. someone calling `stop`?
-                self.thread_event.clear()
-                log.debug("Woke up self.subscription_stream_event_type=%s",
-                         self.subscription_stream_event_type)
-
+                log.debug("Waiting for event")
+                event = self.read_queue.get()
+                log.debug("Woke up event=%s", event)
             log.info("<==")
 
         def poll(self):
             """
-            poll current state according to the subscription_list and add it to stream
+            Poll (invoke SubscriptionEvent.SAMPLE in read) current state
+            according to the subscription_list and add it to request stream.
             """
             log.info("==>")
             # TODO exception
-            self._set_event(self.SubscriptionStreamEventType.GET)
+            self.put_event(self.SubscriptionEvent.SAMPLE)
             log.info("<==")
 
     @abstractmethod
-    def get_subscription_handler(self) -> SubscriptionHandler:
+    def get_subscription_handler(self,
+                                 subscription_list) -> SubscriptionHandler:
         pass
 
     @classmethod
     @abstractmethod
     def get_inst(cls):
         """
         Get adapter instance
+        We use this, since we want to have adapter as singleton.
```

Comment:
Now this is a misleading comment - `get_inst` for the API Adapter returns a new instance (and I'm not really sure why it is needed or if it is a good idea).

Reply:
updated comment